numba_cuda/numba/cuda/cudadrv/driver.py

```diff
@@ -2437,54 +2373,67 @@ def read_func_attr_all(self):
         )
 
 
-def launch_kernel(
-    cufunc_handle,
-    gx,
-    gy,
-    gz,
-    bx,
-    by,
-    bz,
-    sharedmem,
-    hstream,
-    args,
-    cooperative=False,
-):
-    param_ptrs = [addressof(arg) for arg in args]
-    params = (c_void_p * len(param_ptrs))(*param_ptrs)
-
-    params_for_launch = addressof(params)
-    extra = 0
-
-    if cooperative:
-        driver.cuLaunchCooperativeKernel(
-            cufunc_handle,
-            gx,
-            gy,
-            gz,
-            bx,
-            by,
-            bz,
-            sharedmem,
-            hstream,
-            params_for_launch,
-        )
+# Alias for backward compatibility
+CudaPythonFunction = Function
+
+
+def _to_experimental_stream(stream):
+    # stream can be: int (0 for default), Stream (shim), or ExperimentalStream
+    if stream == 0 or stream is None:
+        return ExperimentalStream.from_handle(0)
+    elif isinstance(stream, Stream):
+        # Our shim wraps a cuda.core Stream - just return it!
+        return stream._core_stream
+    elif isinstance(stream, ExperimentalStream):
+        return stream
     else:
-        driver.cuLaunchKernel(
-            cufunc_handle,
-            gx,
-            gy,
-            gz,
-            bx,
-            by,
-            bz,
-            sharedmem,
-            hstream,
-            params_for_launch,
-            extra,
+        raise TypeError(
+            f"Expected a Stream object, ExperimentalStream, or 0, got {type(stream).__name__}"
         )
 
 
+def _normalize_kernel_args(args):
+    # cuda.core.launch can handle ctypes scalars directly, but for pointers
+    # ensure c_void_p is converted to integer (0 if None)
+    return [
+        (arg.value if arg.value is not None else 0)
+        if isinstance(arg, c_void_p)
+        else arg
+        for arg in args
+    ]
+
+
+def _to_cufunction(cufunc_handle):
+    # Ensure handle is a binding.CUfunction
+    if isinstance(cufunc_handle, binding.CUfunction):
+        return cufunc_handle
+    return binding.CUfunction(int(cufunc_handle))
+
+
+# Internal cache for Kernel objects created from CUfunction handles
+_kernel_cache = {}
+
+
+def _get_or_create_kernel(cufunction):
+    key = int(cufunction)
+    kernel = _kernel_cache.get(key)
+    if kernel is not None:
+        return kernel
+    _lazy_init()
+    objcode_stub = object.__new__(ObjectCode)
+    kernel = Kernel._from_obj(cufunction, objcode_stub)
+    _kernel_cache[key] = kernel
+    return kernel
+
+
+def _core_launch(stream, config, cufunction, *args):
+    """
+    Internal helper to launch via cuda.core.experimental.launch ensuring a Kernel object.
+    """
+    kernel = _get_or_create_kernel(_to_cufunction(cufunction))
```

Comment:
I think cufunction should always be a `binding.CUfunction` at this point - if it is not, we should look into what changes are needed to ensure that is true.